src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java

```diff
@@ -0,0 +1,348 @@
+        nullCount += (appendCount - NUMBER_OF_ONES[srcByte & 0xFF]);
+
+        // Sets the bits in destination buffer starting from curDestBitIdx to 0
+        byte destByte = dest.getByte(curDestByteIdx);
+        destByte = (byte) (destByte & ((1 << curDestBitIdx) - 1) & 0xFF);
+
+        // Update destination byte with the bits from source byte
+        destByte = (byte) ((destByte | (srcByte << curDestBitIdx)) & 0xFF);
+        dest.setByte(curDestByteIdx, destByte);
+
+        curSrcBitIdx += appendCount;
+        if (curSrcBitIdx == 8) {
+          curSrcBitIdx = 0;
+          curSrcByteIdx += 1;
+        }
+      } else {
+        // Extract appendCount bits from srcByte, starting from curSrcBitIdx
+        byte mask = (byte) (((1 << leftBitsInCurSrcByte) - 1) & 0xFF);
+        srcByte = (byte) ((srcByte >>> curSrcBitIdx) & mask);
+
+        byte nextSrcByte = src.getByte(curSrcByteIdx + 1);
+        byte nextSrcByteMask = (byte) ((1 << (appendCount - leftBitsInCurSrcByte)) - 1);
+        nextSrcByte = (byte) (nextSrcByte & nextSrcByteMask);
+        nextSrcByte = (byte) (nextSrcByte << leftBitsInCurSrcByte);
+        srcByte = (byte) (srcByte | nextSrcByte);
+
+        nullCount += (appendCount - NUMBER_OF_ONES[srcByte & 0xFF]);
+
+        // Sets the bits in destination buffer starting from curDestBitIdx to 0
+        byte destByte = dest.getByte(curDestByteIdx);
+        destByte = (byte) (destByte & ((1 << curDestBitIdx) - 1));
+
+        // Update destination byte with the bits from source byte
+        destByte = (byte) (destByte | (srcByte << curDestBitIdx));
+        dest.setByte(curDestByteIdx, destByte);
+
+        // Update the source byte index and bit index
+        curSrcByteIdx += 1;
+        curSrcBitIdx = appendCount - leftBitsInCurSrcByte;
+      }
+
+      curIdx += appendCount;
+
+      // Update the destination byte index and bit index
+      curDestBitIdx += appendCount;
+      if (curDestBitIdx == 8) {
+        curDestBitIdx = 0;
+        curDestByteIdx += 1;
+      }
+    }
+
+    return nullCount;
+  }
+
+  private static void appendAllValid(HostMemoryBuffer dest, int startBit, long numRowsLong) {
+    int numRows = toIntExact(numRowsLong);
+    int curDestByteIdx = startBit / 8;
+    int curDestBitIdx = startBit % 8;
+    int curIdx = 0;
+    while (curIdx < numRows) {
+      int leftRowCount = numRows - curIdx;
+      int appendCount;
+      if (curDestBitIdx == 0) {
+        dest.setByte(curDestByteIdx, (byte) 0xFF);
+        appendCount = min(8, leftRowCount);
+      } else {
+        appendCount = min(8 - curDestBitIdx, leftRowCount);
+        byte mask = (byte) (((1 << appendCount) - 1) << curDestBitIdx);
+        byte destByte = dest.getByte(curDestByteIdx);
+        dest.setByte(curDestByteIdx, (byte) (destByte | mask));
+      }
+
+      curDestBitIdx += appendCount;
+      if (curDestBitIdx == 8) {
+        curDestBitIdx = 0;
+        curDestByteIdx += 1;
+      }
+
+      curIdx += appendCount;
+    }
```

Comment:
Same byte-at-a-time and conditionals-in-while-loop performance problems as noted above.
```suggestion
    if (curDestBitIdx > 0) {
      int numBits = 8 - curDestBitIdx;
      int mask = ((1 << numBits) - 1) << curDestBitIdx;
      dest.setByte(curDestByteIdx, (byte) (dest.getByte(curDestByteIdx) | mask));
      curDestByteIdx += 1;
      numRows -= numBits;
    }
    if (numRows > 0) {
      int numBytes = (numRows + 7) / 8;
      dest.setMemory(curDestByteIdx, numBytes, (byte) 0xFF);
    }
```